### PR TITLE
feat: market crud

### PR DESCRIPTION
**Pull request by quasisamurai**
<details>
    <summary><small>resolved</small> problem</summary><p>

Here is yet not finished Proof-of-concept CRUD for Market. While I worked on this, i found some critical issues caused by Solidity & EVM design.  So let's start:

-  I will describe problem using orders, but it's the same for deals. So when we compare two orders (for ex. in OpenDeal) now we're using this construction. Notice, that Order is a struct.
`     Order memory ask = orders[_askID];
        Order memory bid = orders[_bidID]`
This one reffers only for 4 variables, named  

> ask,bid, _askID, _bidID.

But when you're trying to implement CRUD, you need to get exact vars from storage SC. 
There are two noticable things aproaching then:
1. you should use multiple return due solidity currently not supporting struct returning ([only in internal functions](https://solidity.readthedocs.io/en/develop/frequently-asked-questions.html#can-a-contract-function-return-a-struct), that's not applicable for crud)
2. declare every variable and cast a returned value of getter into concrete type, due contract's functions operate with tuples in case of multiple return.

So data taken from storage contract woul'd be look like this:
`( Orders.OrderType orderType,
        address author,
        address counterparty,
        uint duration,
        uint price,
        bool[] memory netflags,
        ProfileRegistry.IdentityLevel identityLevel,
        address blacklist,
        bytes32 tag,
        uint64[] memory benchmarks,
        uint frozenSum) = ord.GetOrderInfo(askID);`   Here is the _part_ of order, the immutable data.

This situation totally **rapes** compiler by calling _only a part of **one** order_. Even without casting this vars as a struct  But we need two full orders. And that's is only begining.

Why this is going on like this? 
Solidity [has](https://ethereum.stackexchange.com/questions/7325/stack-too-deep-try-removing-local-variables) a limit for local variables (16).

> Depending on what you do, you can have around 16 local variables (including parameters and return parameters)

So every input parameter to the function takes one variable, and each return value takes one variable, and each struct cast, and each local declaration takes another variable, and references to storage take two - it surprised me how quickly the limit of 16 was reached.

> struct **a** = **struct**(**b**,**c**,**d**)

will reserve 5 slots in stack (i'm not sure, it's just an empirical conclusion)
 That is an old problem, we faced some months ago and it was a cause why we splitted one order/deal getter by two for both structs.

Since almost each function uses one/two orders or a deal sctruct, there is no way to implement crud  this way. That's not the only issue i faced, but I think this is the most significant problem that needs to be solved.
  </p></details>

So here is implemented crud for market 
Included:
-  Deals

-  Orders

-  ChangeRequests

-  Master-worker contract named Administratum (discussable)

- Also crud for Administratum

- Updated event WorkerConfirmed (api-breaking)

- Updated test enviroment (watch truffle.js & test.sh)

- Aslo fixed some non-critical bugs

By the way we faced and original problem binded with [this](https://github.com/ethereum/EIPs/issues/170) EIP. So we decided to modify our stack to avoid problems with contract's bytecode limit.

p.s. some weird comments and getters will be removed after refactoring iteration soon. :ukraine:

### CODE REVIEW

File: blockchain/source/test/market.js [hunk 1]
```diff
@@ -1281,53 +1286,53 @@ contract('Market', async (accounts) => {
...
-            it('should revert', async () => {
-                await assertRevert(market.RemoveWorker(accounts[0], master, { from: master }));
-            });
-
-            after(async () => {
-                await market.unpause();
-                await market.RemoveWorker(accounts[0], master, { from: master });
-            });
-        });
+        // describe('RegisterWorker', () => {
```
[the remark this answers is not in the document]
Comment (quasisamurai) [reply blockchain/source/test/market.js]: due this section tests cases while market is paused. Now master-worker relations is separated contract.  Later,  me or @abefimov  will update tests and may be administratum would be tested in separate test file. And i leaved (just commented, not deleted) this tests for the future.

File: blockchain/source/contracts/Market.sol [hunk 2]
```diff
@@ -697,41 +535,49 @@ contract Market is Ownable, Pausable {
...
-            bl.Add(deals[dealID].consumerID, deals[dealID].masterID);
+            bl.Add(consumerID, masterID);
         }
     }
 
     function InternalCloseDeal(uint dealID) internal {
-        if (deals[dealID].status == DealStatus.STATUS_CLOSED) {
+        ( , address supplierID, address consumerID, address masterID, , , ) = dl.GetDealInfo(dealID);
+
+        ( , , , Deals.DealStatus status, , , ) = dl.GetDealParams(dealID);
```
[the remark this answers is not in the document]
Comment (quasisamurai) [reply blockchain/source/contracts/Market.sol]: will refactor later, I said.

File: blockchain/source/contracts/Deals.sol [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+
+        return dealsAmount;
+    }
+
+
+
+
+
+    /* function Close(uint dealID) public onlyOwner {
```
[the remark this answers is not in the document]
Comment (quasisamurai) [reply blockchain/source/contracts/Deals.sol]: Will be removed during the refactoring iteration.

File: blockchain/source/contracts/Market.sol [hunk 4]
```diff
@@ -623,72 +454,79 @@ contract Market is Ownable, Pausable {
...
 
-            if (token.balanceOf(deal.consumerID) >= nextPeriodSum) {
-                require(token.transferFrom(deal.consumerID, this, nextPeriodSum));
-                deals[dealID].blockedBalance = deals[dealID].blockedBalance.add(nextPeriodSum);
+            if (token.balanceOf(consumerID) >= nextPeriodSum) {
+                require(token.transferFrom(consumerID, this, nextPeriodSum));
+                dl.SetDealBlockedBalance(dealID, blockedBalance.add(nextPeriodSum));
             } else {
-                emit Billed(dealID, deals[dealID].blockedBalance);
+              // ?????
```
[the remark this answers is not in the document]
Comment (quasisamurai) [reply blockchain/source/contracts/Market.sol]: same.

File: blockchain/source/contracts/Deals.sol [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        uint _duration,
+        uint _price,
+        uint _startTime,
+        uint _endTime,
+        DealStatus _status,
+        uint _blockedBalance,
+        uint _totalPayout,
+        uint _lastBillTS) public onlyOwner returns(uint) {
+
+        dealsAmount += 1;
```
[the remark this answers is not in the document]
Comment (quasisamurai) [reply blockchain/source/contracts/Deals.sol]: Why not? If  a current num of deals is 1, i think it's correct to mark this single deal with index '1'

File: blockchain/source/contracts/Market.sol [hunk 6]
```diff
@@ -120,9 +57,13 @@ contract Market is Ownable, Pausable {
 
     ProfileRegistry pr;
 
-    uint ordersAmount = 0;
+    Orders ord;
+
+    Administratum adm;
 
-    uint dealAmount = 0;
+    Deals dl;
```
[the remark this answers is not in the document]
Comment (quasisamurai) [reply blockchain/source/contracts/Market.sol]: no.

File: blockchain/source/contracts/Market.sol [hunk 7]
```diff
@@ -216,126 +152,161 @@ contract Market is Ownable, Pausable {
...
+        require(!bl.Check(ord.GetOrderBlacklist(_askID), ord.GetOrderAuthor(_bidID)));
+        require(!bl.Check(adm.GetMaster(ord.GetOrderAuthor(_askID)), ord.GetOrderAuthor(_bidID)));
+        require(!bl.Check(ord.GetOrderAuthor(_askID), ord.GetOrderAuthor(_bidID)));
+        require(ord.GetOrderPrice(_askID) <= ord.GetOrderPrice(_bidID));
+        require(ord.GetOrderDuration(_askID) >= ord.GetOrderDuration(_bidID));
         // profile level check
-        require(pr.GetProfileLevel(bid.author) >= ask.identityLevel);
-        require(pr.GetProfileLevel(ask.author) >= bid.identityLevel);
+        require(pr.GetProfileLevel(ord.GetOrderAuthor(_bidID)) >= ord.GetOrderIdentityLevel(_askID));
+        require(pr.GetProfileLevel(adm.GetMaster(ord.GetOrderAuthor(_askID))) >= ord.GetOrderIdentityLevel(_bidID)); //bug
```
Comment (sokel) [inline blockchain/source/contracts/Market.sol]: bug?

File: blockchain/source/contracts/Market.sol [hunk 8]
```diff
@@ -351,266 +322,128 @@ contract Market is Ownable, Pausable {
...
-        return (
-        requests[changeRequestID].dealID,
-        requests[changeRequestID].requestType,
-        requests[changeRequestID].price,
-        requests[changeRequestID].duration,
-        requests[changeRequestID].status
-        );
-    }
-
-    function GetDealsAmount() public view returns (uint){
```
Comment (sokel) [inline blockchain/source/contracts/Market.sol]: it's API breaking, we realy wanna do this?
may be its will be proxy funcs?

File: blockchain/source/contracts/Market.sol [hunk 9]
```diff
@@ -697,41 +537,49 @@ contract Market is Ownable, Pausable {
...
-    }
+
 
     function CalculatePayment(uint _price, uint _period) internal view returns (uint) {
         uint rate = oracle.getCurrentPrice();
         return rate.mul(_price).mul(_period).div(1e18);
     }
 
     function AddToBlacklist(uint dealID, BlacklistPerson role) internal {
+        (, address supplierID, address consumerID, address masterID, , , ) = dl.GetDealInfo(dealID);
```
Comment (sokel) [inline blockchain/source/contracts/Market.sol]: you have getters for this vars;

File: blockchain/source/contracts/Market.sol [hunk 10]
```diff
@@ -787,4 +645,13 @@ contract Market is Ownable, Pausable {
...
         selfdestruct(owner);
     }
+
+    function Migrate(address _newMarket) public onlyOwner returns (bool) {
+        token.transfer(_newMarket, token.balanceOf(address(this)));
+        ord.transferOwnership(_newMarket);
+        dl.transferOwnership(_newMarket);
+        cr.transferOwnership(_newMarket);
+        super.pause();
+
```
Comment (sokel) [inline blockchain/source/contracts/Market.sol]: blank line

File: blockchain/source/migrations/10_set_market_for_blacklist.js [hunk 11]
```diff
@@ -3,13 +3,13 @@ let Blacklist = artifacts.require('./Blacklist.sol');
...
 module.exports = function (deployer, network) {
     if (network === 'private') {
-        Blacklist.deployed()
-            .then(function (blacklist) {
-                blacklist.SetMarketAddress(Market.address, { gasPrice: 0 });
-            })
-            .catch(function (err) {
-                console.log(err);
-            });
+        // Blacklist.deployed()
```
Comment (sokel) [inline blockchain/source/migrations/10_set_market_for_blacklist.js]: why it's commented?

File: blockchain/source/test/market.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sokel) [reply blockchain/source/test/market.js]: keep tests clean.
Remove old and write new.
Don't do it like that.

File: blockchain/source/truffle.js [hunk 12]
```diff
@@ -29,6 +29,8 @@ module.exports = {
             host: 'localhost',
             port: 8535,
             network_id: '*', // eslint-disable-line camelcase
+            gas: 100000000,
+            gasPrice: 1,
```
Comment (sokel) [inline blockchain/source/truffle.js]: why?

File: blockchain/source/truffle.js [hunk 13]
```diff
@@ -58,7 +60,7 @@ module.exports = {
     },
     solc: {
         optimizer: {
-            enabled: true,
+            enabled: false,
```
Comment (sokel) [inline blockchain/source/truffle.js]: ?!?

File: blockchain/source/contracts/Administratum.sol [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+import "zeppelin-solidity/contracts/ownership/Ownable.sol";
+import "./AdministratumCrud.sol";
+
+contract Administratum is Ownable {
+    // events
+
+
+    event WorkerAnnounced(address indexed worker, address indexed master);
+    event WorkerConfirmed(address indexed worker, address indexed master, address indexed confirmator);
```
[the remark this answers is not in the document]
Comment (quasisamurai) [reply blockchain/source/contracts/Administratum.sol]: yup

File: blockchain/source/contracts/Market.sol (hunk 9 above)
Comment (quasisamurai) [reply blockchain/source/contracts/Market.sol]: yes, but I also have one getter for all of them

File: blockchain/source/truffle.js (hunk 13 above)
Comment (quasisamurai) [reply blockchain/source/truffle.js]: fixed

Comment (sokel) [reply blockchain/source/truffle.js]: nope :(

File: blockchain/source/contracts/Administratum.sol [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    //GETTERS
+
+    function GetMaster(address _worker) public view returns (address master) {
+        return  crud.GetMaster(_worker);
+    }
+
+
+
+
+    //modifiers
```
Comment (sokel) [inline blockchain/source/contracts/Administratum.sol]: what is this?

File: blockchain/source/contracts/Administratum.sol [hunk 16]
```diff
@@ -0,0 +1,87 @@
+pragma solidity ^0.4.23;
+
+import "zeppelin-solidity/contracts/ownership/Ownable.sol";
+import "./AdministratumCrud.sol";
+
+contract Administratum is Ownable {
```
Comment (sokel) [inline blockchain/source/contracts/Administratum.sol]: `Administratable` or `Ownable`

**Pull request merged**